File: src/sub_converter/parsers/ssr.py
"""ShadowSocksR (ssr://) parser."""

from __future__ import annotations

import base64
import binascii
from typing import Any

from sub_converter.utils import clean_query, pad_base64


def parse_ssr(url: str) -> dict[str, Any]:
    payload = base64.b64decode(pad_base64(url[len("ssr://"):])).decode("utf-8", errors="strict")
    main_part, _, query_string = payload.partition("/?")
    parts = main_part.split(":")
    if len(parts) < 6:
        raise ValueError("invalid ssr entry")

    server, port, protocol, method, obfs, password_b64 = parts[:6]
    password = base64.b64decode(pad_base64(password_b64)).decode("utf-8", errors="ignore")
    query = clean_query(query_string)

    decoded_query: dict[str, Any] = {}
    for key, value in query.items():
        if key in {"remarks", "group", "protoparam", "obfsparam"}:
            try:
                decoded_query[key] = base64.b64decode(pad_base64(value)).decode("utf-8", errors="ignore")
            except (binascii.Error, ValueError):
                decoded_query[key] = value
        else:
            decoded_query[key] = value

    return {
        "scheme": "ssr",
        "name": decoded_query.get("remarks"),
        "server": server,
        "port": int(port),
        "protocol": protocol,
        "cipher": method,
        "obfs": obfs,
        "password": password,
        "query": decoded_query,
        "raw": url,
    }

File: worker/src/parsers/ssr.ts
/**
 * ShadowSocksR parser
 */

import { ProxyNode } from '../utils/types';
import { padBase64 } from '../services/subscription';
import { cleanQuery } from '../utils/url';

export function parseSSR(entry: string): ProxyNode {
  if (!entry.startsWith('ssr://')) {
    throw new Error('Invalid ssr URL');
  }

  const payload = atob(padBase64(entry.replace('ssr://', '')));

  try {
    const [mainPart, queryString = ''] = payload.split('/?');
    const parts = mainPart.split(':');
    if (parts.length < 6) {
      throw new Error('Invalid ssr format');
    }

    const [server, port, protocol, method, obfs, passwordBase64] = parts;
    const password = atob(padBase64(passwordBase64));
    const query = cleanQuery(queryString);
    const decodedQuery: Record<string, string> = {};
    for (const [key, value] of Object.entries(query)) {
      if (['remarks', 'group', 'protoparam', 'obfsparam'].includes(key)) {
        try {
          decodedQuery[key] = atob(padBase64(value));
        } catch {
          decodedQuery[key] = value;
        }
      } else {
        decodedQuery[key] = value;
      }
    }

    return {
      scheme: 'ssr',
      server,
      port,
      protocol,
      cipher: method,
      obfs,
      password,
      name: decodedQuery.remarks,
      query: decodedQuery,
      raw: entry,
    };
  } catch (e) {
    throw new Error(`Failed to parse ssr: ${e}`);
  }
}
